File: lecture_005/binarysearch.java
import java.util.Scanner;
public class binarysearch{
    public static Scanner scn=new Scanner(System.in);
    public static void main(String[] args){
        solve();
    }

    public static void solve()
    {
        int n=scn.nextInt();
        int[] arr=new int[n];
        input(arr);
        System.out.println("enter number to find");
        int m=scn.nextInt();
        int f=binary(arr,m);
        if(f==-1)
        System.out.println("not found");
        else
        System.out.println("found at  "+f);
        int k=lower(arr,m);
        if(k==-1)
        System.out.println("not found");
        else
        System.out.println("found at  "+k);
        int e=upper(arr,m);
        if(e==-1)
        System.out.println("not found");
        else
        System.out.println("found at  "+e);
    }

     public static void input(int[] arr)
    {
        for(int i=0;i<arr.length;i++)
        arr[i]=scn.nextInt();
        System.out.println();
    }


    public static int binary(int[] arr,int data)
    {
        int ri=arr.length-1,le=0;
        while(le<=ri)
        {
            int mid=(le+ri)/2;
            if(arr[mid]<data)
            le=mid+1;
            else if(arr[mid]>data)
            ri=mid-1;
            else
            return mid;
        }
        return -1;
    
    }

    public static int lower(int[] arr,int data)
    {
        int ri=arr.length-1,le=0;
        while(le<=ri)
        {
            int mid=(le+ri)/2;
            if(arr[mid]<data)
            le=mid+1;
            else if(arr[mid]>data)
            ri=mid-1;
            else if(arr[mid]==data)
            {
                if(arr[mid-1]==data && (mid-1)>=0)
                ri=mid-1;
                else return mid;
                
            }
            else return-1;
            
        }
    
    }
    
    
    public static int upper(int[] arr,int data)
    {
        int ri=arr.length-1,le=0;
        while(le<=ri)
        {
            int mid=(le+ri)/2;
            if(arr[mid]<data)
            le=mid+1;
            else if(arr[mid]>data)
            ri=mid-1;
            else if(arr[mid]==data)
            {
                if(arr[mid+1]==data && (mid+1)!=0)
                le=mid+1;
                else return mid;
            }
            else return -1;
        }
        
    }
    
}

File: lecture_005/binary_search.cpp
#include<iostream>
#include<vector>

using namespace std;

int binary_search(vector<int>& arr, int data){
    int le=0;
    int ri=arr.size()-1;
    int mid=0;
    while(le<=ri){
        mid = (le + ri)/2;
        if(arr[mid]<data)
        le=mid+1;
        else if(arr[mid]>data)
        ri=mid-1;
        else if(arr[mid]==data)
        return mid;
        else return -1;
    }

}

void inputArray(vector<int>& arr){
    for(int i=0;i<arr.size();i++)
        cin>>arr[i];
}

void solve()
{
    int n;
    cin>>n;
    vector<int>arr(n,0);
    inputArray(arr);
    cout<<"number to be searched"<<endl;
    int r;
    cin>>r;
    int m=binary_search(arr,r);
    if(m==-1)
    cout<<"not found";
    else
    {
        cout<<"found at"<<m;
    }
    
}
int main(int args, char**argv)
{
    solve();
    return 0;
}